### DISCUSSION TITLE
FastAPI OIDC+Swagger Example Proposal

### DISCUSSION THREAD

**Discussion by laveolus** [fastapi/fastapi#10175] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
#!/usr/bin/env python3

from typing import Any

from fastapi import FastAPI, HTTPException, Security
from fastapi.openapi.models import OpenIdConnect as OpenIdConnectModel
from fastapi.security import HTTPBearer, SecurityScopes
from fastapi.security.base import SecurityBase
from jwt.api_jwt import PyJWT
from jwt.exceptions import PyJWTError
from jwt.jwks_client import PyJWKClient
from starlette.requests import Request
from starlette.status import HTTP_400_BAD_REQUEST, HTTP_403_FORBIDDEN

# Resource server configuration
jwks_url = "https://dev-00000000.okta.com/oauth2/default/v1/keys"
audience = "api://default"
issuer_url = "https://dev-00000000.okta.com/oauth2/default"

# Swagger configuration (via openapi spec)
oidc_url = (
    "https://dev-00000000.okta.com/oauth2/default/.well-known/openid-configuration"
)

# NOTE, grant types as shown in 'available authorization' modal box in swagger, are determined by the jwks endpoint specifying the grant_types_supported
swagger_ui_init_oauth = {
    "clientId": "00000000000000000000",
    "scopes": [
        "openid",
        "profile",
    ],
    "appName": "Test Application",
    "usePkceWithAuthorizationCodeGrant": True,
}

pyjwt = PyJWT()


class OpenIdConnect(SecurityBase):
    def __init__(self, openIdConnectUrl: str, jwks_url: str, audience: str):
        self.jwks_client = PyJWKClient(jwks_url)

        self.model = OpenIdConnectModel(
            openIdConnectUrl=openIdConnectUrl, description=None
        )
        self.scheme_name = "OpenIdConnect"
        self.audience = audience
        # claim name of the user's oidc groups
        self.scopes_claim = "roles"

    async def __call__(self, security_scopes: SecurityScopes, request: Request) -> Any:
        bearer = await HTTPBearer()(request)
        token = bearer.credentials if bearer else ""
        try:
            signing_key = self.jwks_client.get_signing_key_from_jwt(token)
            data = pyjwt.decode(
                token,
                signing_key.key,
                algorithms=["RS256"],
                audience=self.audience,
                issuer=issuer_url,
            )
        except PyJWTError:
            raise HTTPException(
                status_code=HTTP_400_BAD_REQUEST,
                detail="Unsupported authorization code",
            )

        if security_scopes.scopes:
            if not set(security_scopes.scopes).issubset(set(data[self.scopes_claim])):
                raise HTTPException(
                    status_code=HTTP_403_FORBIDDEN, detail="Not Authorized"
                )
        return data


openid_connect = OpenIdConnect(oidc_url, jwks_url, audience)

app = FastAPI(swagger_ui_init_oauth=swagger_ui_init_oauth)


@app.get("/test_foo")
async def test_foo(token: Any = Security(openid_connect, scopes=["Foo Group"])):
    return token


@app.get("/test_bar")
async def test_bar(token: Any = Security(openid_connect, scopes=["Bar group"])):
    return token


@app.get("/test_everyone_and_foo")
async def test_everyone_and_foo(
    token: Any = Security(openid_connect, scopes=["Everyone", "Foo Group"])
):
    return token
```

### Intent
My goal was to create a simple example how to use OIDC authorization code flow (w/ PKCE), using FastAPI as resource server, including the swagger UI.

I'm not an expert in these matters and the documentation I found so far was lacking (and the crucial difference is a bit subtle), so I have made a working example that may help to explain that.

The main difference of OIDC compared to OAuth2, is that the authorization scope needs to be mapped _manually_, as this is not standardized. Typically the authorization server injects group names of the user in a custom claim of the access token, which can then be validated by the resource server for authorization.

What do I need to do to either get this example as documentation in mainline, and/or to get my OpenIdConnect implementation in mainline?

### Request for comments

Even though it works (for Okta at least), I think I need advise to make this code maintainable,

* Use of HTTPBearer seems dodgy, but if I would use it as a nonfield param it would automatically be registered in the openapi schema.
* I'm _reusing_ the Security nonfield param of the oauth 2 scheme to map authorization claims. It works technically, but I'm not sure whether this use fits the design as it isn't a 'scope' (i.e. the scp claim).
* I'm not sure the PyJWT library is the best dependency for FastAPI. Jose does not fetch/cache the jwks resource, so I used  PyJWT instead. So far I could not find the sweet spot of a working and nonscary auth library: reading the FastAPI oauth2 docu, jose works, but requires a scary amount of code to maintain and other fastapi auth libraries (okta_jwt and fastapi-third-party-auth) were not scary, but did not work for my use case.
* The swagger UI is very bloated in providing all grant_types_supported options and all scopes. I'm not sure this can be filtered somehow as most of them are irrelevant and/or do not work.

### Operating System

Linux

### FastAPI Version

0.103.0

### Pydantic Version

2.3.0

### Python Version

3.10.12

**Comment by dancaugherty** [comment]
+1 !

The documentation for authentication code flows with PKCE does not exist. This is very odd when one considers that major third party vendors are pushing everyone to abandon the implicit flow in favor of this flow. 

At this point, I'm not convinced anyone really knows how to implement it.

**Comment by jtele2** [comment]
I am very confused on how to implement OIDC in FastAPI as well. You're example is essentially what I am doing - instead of defining a new `OpenIdConnect` class though, I'm using a dependency function that takes in the current `fastapi.security.OpenIdConnect` and does exactly what you do here. To me, what you have written is what the class should have always been!

**Comment by adrienaggery** [comment]
I need to implement jwt verification and after some head scratches it looks like the OpenIdConnect class in fastapi.security does nothing appart from extracting any string from the Bearer header and returning it (or throwing 40x if not present).

Is it me or this class should actually be implemented to something like what @laveolus did ?
I mean getting kid, getting public keys from OIDC and actually decoding the token before returning it to the handler.

I think this is something important that should be pushed to be implemented like this.

**Comment by laveolus** [comment]
Unfortunately there does not seem a standard way to "just implement OIDC", as every OIDC provider and its users may map claims differently and therefore need to be verified in a specific way in the resource server and/or requested from your SPA. Also they may use some nonstandard extensions to be more secure (and/or create more lock-in).

I see there is a library that implements Zitadel OIDC associated to a pull request I made for a tutorial, that seems well-built, but specific to Zitadel. If you prefer, you might want to fork/extend it for your implementation, as this is seems a good basis:
https://github.com/fastapi/fastapi/pull/10278#issuecomment-2690212858
